### ISSUE TITLE
Istio Telemetry: jaeger goroutines stuck on generating a random number

### ISSUE THREAD

**Issue by mandarjog** [istio/istio#10258]
Istio version 1.0.4
Under heavy load jaeger startspan function is called from 1000s of go routines.
They are attempting to start tracing and contend for a single mutex to generate a random number.

There is a different change that removing random tracing from mixer would address this problem. However if we want to increasing tracing in Mixer we are bound to encounter this problem again.

```
goroutine 38746 [semacquire]: 2051 times: 
sync.runtime_SemacquireMutex(0xc0009be5a4, 0xc00e71e700)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0009be5a0)
	/usr/local/go/src/sync/mutex.go:134 +0xff
istio.io/istio/vendor/github.com/uber/jaeger-client-go/utils.(*lockedSource).Int63(0xc0009be5a0, 0xc3e4316820)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/github.com/uber/jaeger-client-go/utils/rand.go:35 +0x2d
math/rand.(*Rand).Int63(0xc00045d590, 0x0)
	/usr/local/go/src/math/rand/rand.go:85 +0x33
istio.io/istio/vendor/github.com/uber/jaeger-client-go.NewTracer.func2(0x27fe955994)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/github.com/uber/jaeger-client-go/tracer.go:124 +0x2a
istio.io/istio/vendor/github.com/uber/jaeger-client-go.(*Tracer).randomID(0xc000280c80, 0x0)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/github.com/uber/jaeger-client-go/tracer.go:404 +0x2e
istio.io/istio/vendor/github.com/uber/jaeger-client-go.(*Tracer).startSpanWithOptions(0xc000280c80, 0xc3eb012260, 0x10, 0xc3f6b00720, 0x1, 0x1, 0xbef988590e6b1626, 0x27fe955994, 0x2d57420, 0x0, ...)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/github.com/uber/jaeger-client-go/tracer.go:252 +0x962
istio.io/istio/vendor/github.com/uber/jaeger-client-go.(*Tracer).StartSpan(0xc000280c80, 0xc3eb012260, 0x10, 0xc3f712a3f0, 0x1, 0x1, 0x0, 0x1)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/github.com/uber/jaeger-client-go/tracer.go:184 +0xf0
istio.io/istio/vendor/github.com/opentracing/opentracing-go.startSpanFromContextWithTracer(0x1d523c0, 0xc11abf3d70, 0x1d3fdc0, 0xc000280c80, 0xc3eb012260, 0x10, 0x0, 0x0, 0x0, 0x10, ...)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/github.com/opentracing/opentracing-go/gocontext.go:52 +0x127
istio.io/istio/vendor/github.com/opentracing/opentracing-go.StartSpanFromContext(0x1d523c0, 0xc11abf3d70, 0xc3eb012260, 0x10, 0x0, 0x0, 0x0, 0x0, 0xc3c0d173b0, 0x1d523c0, ...)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/github.com/opentracing/opentracing-go/gocontext.go:44 +0x95
istio.io/istio/mixer/pkg/api.(*grpcServer).Report(0xc000d30ff0, 0x1d523c0, 0xc11abf3c80, 0xc1454b7900, 0xc000d30ff0, 0x0, 0xc0558656e8)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/mixer/pkg/api/grpcServer.go:255 +0x436
istio.io/istio/vendor/istio.io/api/mixer/v1._Mixer_Report_Handler.func1(0x1d523c0, 0xc11abf3c80, 0x19b3280, 0xc1454b7900, 0xc00e576f40, 0x1ae9baa, 0x5, 0x1b0d5d6)
	/Volumes/Untitled/mjog/GOHOME/src/istio.io/istio/vendor/istio.io/api/mixer/v1/service.pb.go:113 +0x86

```

**Comment by yurishkuro** [comment]
Yes, the random number generator in Jaeger is synchronized: https://github.com/jaegertracing/jaeger-client-go/blob/c58709fe19d793cdd40ad8bfa6e645a40a2eccf8/utils/rand.go#L35

One option might be to have a pool of these - it would still require a mutex, but hopefully with a lot less contention.

**Comment by douglas-reid** [comment]
@objectiser @yurishkuro is there any planned work on the Jaeger side in the near term to address the options listed here and in the issue in the Jaeger repo?

The reason I'm asking: If not, we need to start building a workaround into our processes for an upcoming release. Given a few pointers, perhaps we could help?

Thanks for any advice you can provide.

**Comment by yurishkuro** [comment]
@douglas-reid help with fixing it in the client ([issue #353](https://github.com/jaegertracing/jaeger-client-go/issues/353)) would be most appreciated.  We don't exactly have a solution to this, I proposed above using a pool of random number generators, but it would be interesting to observe directly in Istio whether that even addresses your contention.

**Comment by objectiser** [comment]
@mandarjog Updated commit: https://github.com/jaegertracing/jaeger-client-go/commit/3e70059a3935599fa849da15dddde30cde5a6226

**Comment by mandarjog** [comment]
Thanks @objectiser 
We recently started using `jaeger.NewProbabilisticSampler` with `0` random sampling.
My expectation was that this will reduce the load on this call, but that has not happened.
`startSpanWithOptions` calls `t.randomID()` before consulting if a new trace should be originated.
This means that even with `0%` tracing we end up generating all these IDs'

Can you confirm this @objectiser ?
I will have results of our experiment soon.

**Comment by yurishkuro** [comment]
Sampling and generating trace/span IDs are not coupled in any way. There are many good reasons to have proper IDs even when sampling is off.

**Comment by mandarjog** [comment]
A Lot of tracing machinery is exercised even though the present request is not traced, it involves path length and allocs. What are some other reasons why this work needs to be done when the request is not going to be traced?

**Comment by douglas-reid** [comment]
@yurishkuro just curious: as Mixer operates almost as a meta-service, if we wanted to avoid ID generation when the request is not sampled, would there be harm in somehow modifying the OT middleware we employ to return, essentially, a `noopSpan`-like struct from the `tracer.StartSpan` call in the interceptor ?

**Comment by objectiser** [comment]
@yurishkuro Agree with your examples.

@mandarjog @douglas-reid At present mixer is a leaf service, so doesn't propagate the trace context any further. Is this possibly going to change in the future? e.g. tracing into adapters?

If not, then might be reasonable to have a noopSpan returned.

**Comment by douglas-reid** [comment]
This should be addressed by https://github.com/istio/istio/pull/10826.

**Comment by munrodg** [comment]
Closed due to #10826 merge.
